src/main/java/com/google/devtools/build/lib/collect/nestedset/Depset.java

```diff
@@ -374,10 +376,22 @@ public static Depset fromDirectAndTransitive(
               order.getStarlarkName(), x.getOrder().getStarlarkName());
         }
         builder.addTransitive(x.getSet());
+        nonEmptyTransitives++;
       }
     }
 
-    return new Depset(type, builder.build());
+    NestedSet<Object> set = builder.build();
+    // If the nested set was optimized to the only non-empty transitive element, reuse the
+    // corresponding depset.
+    if (nonEmptyTransitives == 1) {
```

Comment:
@comius I could see an argument for removing this check and just always checking whether the set just built equals any of the transitive sets. What do you think?

Reply:
The implementation makes the reader think, it's dropping `direct` sets. The `nonEmptyTransitives` is the misdirection that makes you look into Nestedset implementation.

The argument why this works, without looking into the implementation is:
"If NestedSetBuilder.build returned one of the transitive Nestedset, so can this code return one of the transitive Depsets"

Can you change and document the code to follow this premise?

That is, remove the `nonEmptyTransitives` variable and just do the search. I'd find this easier to read.

Reply:
Changed.